Move implementation of DISTINCT aggregates to optimizer rule

martint opened the pull request:
The planner now does a more straightforward translation of DISTINCT
aggregates. Their implementation (either by using MarkDistinct or
a pre-aggregation) is done as optimizer rules.

Also, fixes #10156

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementSingleDistinctAggregation.java:
```diff
@@ -0,0 +1,148 @@
+
+        Set<Set<Expression>> uniqueArgumentSet = ImmutableSet.copyOf(argumentSets);
+
+        if (uniqueArgumentSet.size() != 1 || argumentSets.size() != aggregations.size()) {
```
nit: I think you could move this condition to pattern matching and catch `Iterables.getOnlyElement(uniqueArgumentSet)` as capture

sopel39 replied:
I don't think `uniqueArgumentSet.size() != 1` is needed at this point as pattern already enforces that. You could do check state though.

martint replied:
Will do. I just copied this from the old optimizer

martint replied:
> and catch Iterables.getOnlyElement(uniqueArgumentSet) as capture

Are there any examples on how to capture those values in the pattern match? All I see are usages where nodes are captured.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementSingleDistinctAggregation.java:
```diff
@@ -0,0 +1,148 @@
+                        new AggregationNode(
+                                context.getIdAllocator().getNextId(),
+                                aggregation.getSource(),
+                                Collections.emptyMap(),
```
use `ImmutableMap.of()`, because it will be copied to it anyway

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementMultipleDistinctAggregation.java:
```diff
@@ -0,0 +1,140 @@
+ *              - X
+ * </pre>
+ */
+public class ImplementMultipleDistinctAggregation
```
no unit tests?

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementSingleDistinctAggregation.java:
```diff
@@ -0,0 +1,148 @@
+ * <p>
+ * Assumes s0, s1, ... are symbol references (i.e., complex expressions have been pre-projected)
+ */
+public class ImplementSingleDistinctAggregation
```
no unit tests?

sopel39 replied:
Remove `SingleMarkDistinctToGroupBy`?

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementMultipleDistinctAggregation.java:
```diff
@@ -0,0 +1,140 @@
+    public Result apply(AggregationNode parent, Captures captures, Context context)
+    {
+        // the distinct marker for the given set of input columns
+        Map<Set<Symbol>, Symbol> markers = new HashMap<>();
```
Side note: I would add check state to `AggregationNode` that all aggregation `FunctionCall` have only symbol expressions

martint replied:
That should be caught by:

```
Set<Symbol> inputs = call.getArguments().stream()
        .map(Symbol::from)
        .collect(Collectors.toSet());
```

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementMultipleDistinctAggregation.java:
```diff
@@ -0,0 +1,140 @@
+
+                Symbol marker = markers.get(inputs);
+                if (marker == null) {
+                    marker = context.getSymbolAllocator().newSymbol(inputs.iterator().next().getName(), BOOLEAN, "distinct");
```
just use: https://google.github.io/guava/releases/21.0/api/docs/com/google/common/collect/Iterables.html#getLast-java.lang.Iterable- instead of `inputs.iterator().next().getName()`
?

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementMultipleDistinctAggregation.java:
```diff
@@ -0,0 +1,140 @@
+                        .map(Symbol::from)
+                        .collect(Collectors.toSet());
+
+                Symbol marker = markers.get(inputs);
```
use https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#computeIfAbsent-K-java.util.function.Function-
?

martint replied:
I can't. The body of the `if` needs to create the MarkDistinctNode

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementMultipleDistinctAggregation.java:
```diff
@@ -0,0 +1,140 @@
+                    markers.put(inputs, marker);
+                }
+
+                ImmutableSet.Builder<Symbol> distinctSymbols = ImmutableSet.<Symbol>builder()
```
maybe extract:
```
Set<Symbol> commonDistinctSymbols = ImmutableSet.<Symbol>builder()
  .addAll(parent.getGroupingKeys())
  .build();
parent.getGroupIdSymbol().ifPresent(distinctSymbols::add);
```
so that code is more obvious.

martint replied:
Doesn't seem to stand independently. That's just computing the columns the MarkDistinct needs to evaluate. I'll add a comment, though.

martint replied:
Although, it's a lot more obvious when I place it next to the creation of MarkDistinct

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementMultipleDistinctAggregation.java:
```diff
@@ -0,0 +1,140 @@
+                                aggregation.getSignature(),
+                                Optional.of(marker)));
+
+                subPlan = new MarkDistinctNode(
```
+1. You could also build `MarkDistinctNodes` from `Map<Set<Symbol>, Symbol> markers = new HashMap<>();` after you pass over all aggregations.

> Alternatively, maybe we could create it always (and maybe not even cache markers here). It could be solved by another rule which would find a cascade of MarkDistinctNodes which have the same distinctSymbols so they could be optimized.

Good idea. They don't also have to have the same `distinctSymbols` but rather widening `distinctSymbols`, e.g: `[a] -> [a, b] -> [a, b, c]`...
It seems that finding optimal order might be cost decision. Is it NP hard problem?
Consider symbol sets: `[a], [a, b], [a, c]`. You have two options:
```
1. [a], [a,b], [a, c]
2. [a], [a,c], [a, b]
```

martint replied:
Oops.. That’s what I meant to do. In a previous version it was iterating over the markers instead of the aggregations

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:
```diff
@@ -253,6 +254,8 @@ public PlanOptimizers(
                                         new PushLimitThroughSemiJoin(),
                                         new RemoveTrivialFilters(),
                                         new ImplementFilteredAggregations(),
+                                        new ImplementSingleDistinctAggregation(),
```
Order between:
```
new ImplementSingleDistinctAggregation(),
new ImplementMultipleDistinctAggregation(),
```
is important here. However, `IterativeOptimizer` doesn't give you (explicitly) any guarantees.
Consider putting those rules in separate `IterativeOptimizer` executions or make `ImplementMultipleDistinctAggregation` not execute on `single distinct aggregation`

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementSingleDistinctAggregation.java:
```diff
@@ -0,0 +1,148 @@
+
+    private static boolean hasDistinct(AggregationNode aggregation)
+    {
+        return aggregation.getAggregations()
```
What if aggregation has filters. See:
`com/facebook/presto/sql/planner/iterative/rule/SingleMarkDistinctToGroupBy.java:62`

martint replied:
Good catch. I'll add a filter for that.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementSingleDistinctAggregation.java:
```diff
@@ -0,0 +1,148 @@
+        return aggregation.getAggregations()
+                .values().stream()
+                .filter(e -> e.getCall().isDistinct())
+                .map(e -> e.getCall().getArguments())
```
you miss `.map(HashSet::new)` as in https://github.com/prestodb/presto/pull/10157/commits/739ec8219a4b9f605bde68080921c21cbbefc006#diff-dbe409204f565f57244ce6146995bdbeR93.
The rule won't fire even if it could in case of redundant symbols

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementSingleDistinctAggregation.java:
```diff
@@ -0,0 +1,148 @@
+    {
+        return aggregation.getAggregations()
+                .values().stream()
+                .filter(e -> e.getCall().isDistinct())
```
Add check argument to `com.facebook.presto.sql.planner.plan.AggregationNode.Aggregation` that either function call is distinct or marker is set, but not both.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementMultipleDistinctAggregation.java:
```diff
@@ -0,0 +1,140 @@
+    {
+        return aggregation.getAggregations()
+                .values().stream()
+                .anyMatch(e -> e.getCall().isDistinct() && !e.getMask().isPresent());  // MarkDistinct doesn't support a mask (yet), so skip if aggregation has mask set
```
This is for the pattern. That case is caught by the analyzer (I think) and, ultimately, by a planvalidator once optimization is done.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementSingleDistinctAggregation.java:
```diff
@@ -0,0 +1,148 @@
+ * <p>
+ * Assumes s0, s1, ... are symbol references (i.e., complex expressions have been pre-projected)
+ */
+public class ImplementSingleDistinctAggregation
```
It shouldn’t matter, but I need to make sure the multiple distinct rule doesn’t fire for the single distinct case.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java:
```diff
@@ -348,18 +347,6 @@ public FeaturesConfig setOptimizeHashGeneration(boolean optimizeHashGeneration)
         return this;
     }
 
-    public boolean isOptimizeSingleDistinct()
```
It's been the default for 3 years and we've never had to use the other variant in this case.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementSingleDistinctAggregation.java:
```diff
@@ -0,0 +1,148 @@
+                        false,
+                        call.getArguments()),
+                aggregation.getSignature(),
+                Optional.empty());
```
That came from the old optimizer, but I'll fix it.

martint commented:
Updated. Added some tests, but still missing tests for the Multiple->MarkDistinct rule. The matchers for aggregations with masks seem to be missing some features (or I just couldn't get it to work after a couple of attempts -- I left a commented-out test)

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MultipleDistinctAggregationToMarkDistinct.java:
```diff
@@ -0,0 +1,160 @@
+                .filter(FunctionCall::isDistinct)
+                .count();
+
+        return distincts > 0 && distincts < aggregation.getAggregations().size();
```
why non distinct aggregation is required? Javadoc contradicts this predicate:
```
+ * Implements distinct aggregations with different inputs by transforming plans of the following shape:
+ * <pre>
+ * - Aggregation
+ *        GROUP BY (k),
+ *        F1(DISTINCT s0, s1, ...),
+ *        F2(DISTINCT s2, s3, ...)
```

martint replied:
This handles the case where there’s at least one non-distinct agg. That is not something the other rule (single distinct to group by) can deal with

sopel39 replied:
> single distinct to group by

but for single distinct to group by all aggregation must have same symbols.

martint replied:
Yes, so:

* if there are more than one different distincts OR
* there’s at least one non-distinct

This rule will handle it. I can update the example to reflect that

sopel39 replied:
ah right. Haven't seen `or` there. My bad.

sopel39 commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestMultipleDistinctAggregationToMarkDistinct.java:
```diff
@@ -0,0 +1,121 @@
+                        .addAggregation(p.symbol("output2"), expression("sum(DISTINCT input)"), ImmutableList.of(BIGINT))
+                        .source(
+                                p.values(p.symbol("input")))))
+                .doesNotFire();
```
why so?

martint replied:
This is handled by single distinct to group by rule. It verified that this rule will not try to rewrite the plan in such case to avoid rule application order issues

martint replied:
This is handled by single distinct to group by rule. It verified that this rule will not try to rewrite the plan in such case to avoid rule application order issues

sopel39 commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestMultipleDistinctAggregationToMarkDistinct.java:
```diff
@@ -0,0 +1,121 @@
+    @Test
+    public void testMixedDistinctAndNonDistinct()
+    {
+//        tester().assertThat(new MultipleDistinctAggregationToMarkDistinct())
```
uncomment?

sopel39 replied:
I see your comment. Will you still work on that?

martint replied:
I'm don't think I'll be able to get to that in the near future. I'll leave those tests for later. There's enough coverage with the end-to-end tests I added in this commit and to ensure the rule doesn't fire when it shouldn't.

sopel39 replied:
> There's enough coverage with the end-to-end tests I added in this commit and to ensure the rule doesn't fire when it shouldn't.

Right, but was it rewritten previously? Could not rewiring mixed dist aggregation have performance impact of some queries?

Do you consider this PR done?

martint replied:
Not sure I understand your question. If it’s not rewritten, the query will fail because the operators cannot deal with aggregations that have a DISTINCT in them.

Yes, other than those missing tests, this should be done now.

sopel39 replied:
> If it’s not rewritten, the query will fail because the operators cannot deal with aggregations that have a DISTINCT in them.

What rewrites mixed aggregations with different distinct sets after your changes? e.g:
```count(distinct a), count(distinct b), count(c)```.

I'm probably missing some context here.

martint replied:
Originally, this was all handled during initial SQL -> IR translation. Now, all that logic is in this rule. What I meant above is that if this rule doesn't catch something, it will just fail with a planning error. It won't affect performance.

sopel39 commented:
> The matchers for aggregations with masks seem to be missing some features 

If aggregation has mask does that mean it was already processed by your rule?

martint commented:
> If aggregation has mask does that mean it was already processed by your rule?

Not necessarily. Filtered aggregates are implemented in the same way. BTW, I was referring to the matchers for the asserting the shape of the transformed plan, not the rule precondition

sopel39 commented on presto-main/src/test/java/com/facebook/presto/sql/query/TestDistinctAggregations.java:
```diff
@@ -0,0 +1,212 @@
+                "VALUES (REAL '1.0', REAL '1.0', BIGINT '4')");
+
+        assertions.assertQuery(
+                "SELECT corr(DISTINCT x, y), corr(DISTINCT y, x), count(DISTINCT x) FROM " +
```
Add test (mixed aggregation with different symbol sets)?
```
SELECT count(DISTINCT x), count(DISTINCT y), count(z) FROM
```

sopel39 commented on presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestMultipleDistinctAggregationToMarkDistinct.java:
```diff
@@ -0,0 +1,121 @@
+    @Test
+    public void testMixedDistinctAndNonDistinct()
+    {
+//        tester().assertThat(new MultipleDistinctAggregationToMarkDistinct())
```
please add TODO/comment why it's commented out

martint replied:
I'll just remove it and put it in a github issue for later. I had it here for illustration and in case someone had ideas about something I might be missing.

martint commented:
updated
